### PR TITLE
【Hackathon 6th No.31】paddle.distribution.Normal support complex normal distribution -part

### PR DESCRIPTION
**Pull request by NKNaN**
### PR Category

User Experience

### PR Types

New features

### Description

Upgarde paddle.distribution.Normal to support complex normal distribution

### CODE REVIEW

File: python/paddle/distribution/normal.py [hunk 1]
```diff
@@ -40,14 +40,24 @@ class Normal(distribution.Distribution):
...
-    In the above equation:
+    If 'loc' is complex number, the probability density function (pdf) is
+
+    .. math::
+
+        pdf(x; \mu, \sigma) = \frac{1}{Z}e^{\frac {-(x - \mu)^2}  {\sigma^2} }
+
+    .. math::
+
+        Z = \pi \sigma^2
```
Comment (zhiminzhang0830) [inline python/paddle/distribution/normal.py:51]: 这里没有根号么

File: python/paddle/distribution/normal.py [hunk 2]
```diff
@@ -128,33 +139,84 @@ def __init__(self, loc, scale, name=None):
...
+            loc = np.array(loc)
+            if loc.dtype == np.float64:
+                loc = loc.astype('float32')
+            if loc.dtype == np.complex128:
+                loc = loc.astype('complex64')
+
+        if isinstance(scale, (tuple, list)):
+            scale = np.array(scale)
+            if scale.dtype == np.float64:
+                scale = scale.astype('float32')
```
Comment (zhiminzhang0830) [inline python/paddle/distribution/normal.py]: 这种情况下为啥需要转换成float32呢？

File: python/paddle/distribution/normal.py [hunk 3]
```diff
@@ -234,18 +304,26 @@ def rsample(self, shape=()):
...
 
         .. math::
 
             entropy(\sigma) = 0.5 \log (2 \pi e \sigma^2)
 
+        If complex gaussian, the entropy is
+
+        .. math::
+
+            entropy(\sigma) = \log (\pi e \sigma^2) + 1
```
Comment (zhiminzhang0830) [inline python/paddle/distribution/normal.py:323]: 辛苦确认公式的正确性，最好贴一下推导过程

File: python/paddle/distribution/normal.py (hunk 1 above)
Comment (NKNaN) [reply python/paddle/distribution/normal.py:51]: 应该没有的，
若 $Z = X + iY$ ，且 $X \sim N(\mu, \sigma^2)$ , $Y \sim N(\mu, \sigma^2)$ ， $X$ 与 $Y$ 相互独立，则 $Z \sim CN(\mu+i\mu, 2\sigma^2)$  
记 $\mu+i\mu = \mu_z \in \mathbb{C}$ ， $2\sigma^2 = \sigma_z^2 \in \mathbb{R}$ ，则其概率密度函数为：

$$ \begin{aligned} p_Z(z) = p_{xy}(x, y) & = \frac{1}{\sqrt{2 \pi \sigma^2}} \exp[-\frac{(x-\mu)^2}{2 \sigma^2}] \cdot \frac{1}{\sqrt{2 \pi \sigma^2}} \exp[-\frac{(y-\mu)^2}{2 \sigma^2}] \\
& = \frac{1}{\pi \cdot 2 \sigma^2} \exp[-\frac{(x-\mu)^2+(y-\mu)^2}{2 \sigma^2}] \\
& = \frac{1}{\pi \cdot 2 \sigma^2} \exp[-\frac{|(x-\mu)+(y-\mu) \cdot i |^2}{2 \sigma^2}] \\
& = \frac{1}{\pi \cdot 2 \sigma^2} \exp[-\frac{|(x+y \cdot i)-(\mu+\mu \cdot i) |^2}{2 \sigma^2}] \\
& =\frac{1}{\pi\sigma_z^2}\exp[-\frac{|z - \mu_z|^2}{\sigma_z^2}] \end{aligned}$$

File: python/paddle/distribution/normal.py (hunk 2 above)
Comment (NKNaN) [reply python/paddle/distribution/normal.py]: 这里是想把 tuple 和 list 先转成 np.adarray，因为numpy的 np.array 默认是 np.float64，这里是转成paddle的默认类型。是不是这样更好一点：
if isinstance(scale, (tuple, list)):
    scale = np.array(scale, dtype=np.float32)

File: python/paddle/distribution/normal.py (hunk 3 above)
Comment (NKNaN) [reply python/paddle/distribution/normal.py:323]: Entopy:

$$ \begin{aligned}
H(Z) & = - \int_z p(z)\log p(z) dz \\
& = - \mathbb{E}[\log p(z)] \\
& = -  \mathbb{E}[-\log(\pi \sigma_z^2)-\frac{1}{\sigma_z^2}|z-\mu_z|^2] \\
& = \log(\pi \sigma_z^2) + \frac{1}{\sigma_z^2}\mathbb{E}[|z-\mu_z|^2] \\
& = \log(\pi \sigma_z^2) + 1
\end{aligned}$$

KL-DIV:

$$ \begin{aligned}
\mathcal{D}\_{KL}(p || q) & = \int_z p(z)\log \frac{p(z)}{q(z)} dz \\
& = \mathbb{E}\_p\[\log \frac{p(z)}{q(z)}\] \\
& = \mathbb{E}\_p\[\log\[\frac{\sigma_q^2}{\sigma_p^2} \exp(-\frac{|z-\mu_p|^2}{\sigma_p^2}+\frac{|z-\mu_q|^2}{\sigma_q^2})\]\] \\
& = -\log \frac{\sigma_p^2}{\sigma_q^2} - \frac{1}{\sigma_p^2}\mathbb{E}\_p(z-\mu_p)^2 + \frac{1}{\sigma_q^2}\mathbb{E}\_p(z-\mu_q)^2\\
& = -2\log \frac{\sigma_p}{\sigma_q} - 1 + \frac{1}{\sigma_q^2}\mathbb{E}\_p(z^cz -\mu_q^cz-z^c\mu_q+\mu_q^c\mu_q) \\
& = -2\log \frac{\sigma_p}{\sigma_q} - 1 + \frac{1}{\sigma_q^2}(\sigma_p^2+\mu_p^c\mu_p -\mu_q^c\mu_p-\mu_p^c\mu_q+\mu_q^c\mu_q) \\
& = -2\log \frac{\sigma_p}{\sigma_q} - 1 + \frac{1}{\sigma_q^2}(\sigma_p^2+|\mu_p-\mu_q|^2) \\
& = \frac{\sigma_p^2}{\sigma_q^2} - 1 + \frac{|\mu_p-\mu_q|^2}{\sigma_q^2}-2\log \frac{\sigma_p}{\sigma_q}
\end{aligned} $$

File: python/paddle/distribution/normal.py (hunk 2 above)
Comment (zhiminzhang0830) [reply python/paddle/distribution/normal.py]: 嗯嗯，改成这样把

Comment (NKNaN) [reply python/paddle/distribution/normal.py]: 已修改

File: test/distribution/test_distribution_normal.py [hunk 4]
```diff
@@ -268,9 +301,33 @@ def test_normal_distribution_static(self, sample_shape=7, tolerance=1e-6):
...
+        # loc is 'complex' and scale is 'float'
+        m = int((np.random.ranf() - 0.5) * 8)
+        self.loc_np = m + m * 1j
+        self.scale_np = (np.random.ranf() - 0.5) * 4
+        while self.scale_np < 0:
+            self.scale_np = (np.random.ranf() - 0.5) * 4
+        # used to construct another Normal object to calculate kl_divergence
+        m2 = int((np.random.ranf() - 0.5) * 8)
+        self.other_loc_np = m2 + m2 * 1j
+        self.other_scale_np = int((np.random.ranf() - 0.5) * 8)
```
Comment (zhiminzhang0830) [inline test/distribution/test_distribution_normal.py]: 这里为啥scale_np和other_scale_np后面的系数不一样

File: test/distribution/test_distribution_normal.py [hunk 5]
```diff
@@ -315,6 +396,42 @@ def init_static_data(self, batch_size, dims, in_pir=False):
...
+        v1 = np.random.randn(batch_size, dims)
+        v2 = np.random.randn(batch_size, dims)
+        self.values_np = np.vectorize(complex)(v1, v2).astype('complex64')
+        # used to construct another Normal object to calculate kl_divergence
+        m2 = (np.random.ranf() - 0.5) * 4
+        self.other_loc_np = m2 + m2 * 1j
+        self.other_scale_np = np.random.randn(batch_size, dims).astype(
+            'float32'
+        )
+        while not np.all(self.scale_np > 0):
```
Comment (zhiminzhang0830) [inline test/distribution/test_distribution_normal.py]: 这里是不是应该是判断other_scale_np

Comment (NKNaN) [reply test/distribution/test_distribution_normal.py]: 应该是的，已修改

File: test/distribution/test_distribution_normal.py (hunk 4 above)
Comment (NKNaN) [reply test/distribution/test_distribution_normal.py]: 已修改为一致

**Review by zhiminzhang0830 — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Comment by luotao1** [comment]
请更新对应的中文文档

**Pull request merged**